pkg/operator/apiserver/controller/apiservice/apiservice_controller.go

```diff
@@ -73,53 +80,57 @@ func (c *APIServiceController) sync(ctx context.Context, syncCtx factory.SyncCon
 	case operatorsv1.Unmanaged:
 		return nil
 	case operatorsv1.Removed:
-		errs := []error{}
-		apiServices, err := c.getAPIServicesToManageFn()
+		enabledApiServices, disabledApiServices, err := c.getAPIServicesToManageFn()
 		if err != nil {
-			errs = append(errs, err)
-			return errors.NewAggregate(errs)
-		}
-		for _, apiService := range apiServices {
-			if err := c.apiregistrationv1Client.APIServices().Delete(ctx, apiService.Name, metav1.DeleteOptions{}); err != nil {
-				errs = append(errs, err)
-			}
+			return err
 		}
-		return errors.NewAggregate(errs)
+		return c.syncDisabledAPIServices(ctx, append(enabledApiServices, disabledApiServices...))
 	default:
 		syncCtx.Recorder().Warningf("ManagementStateUnknown", "Unrecognized operator management state %q", operatorConfigSpec.ManagementState)
 		return nil
 	}
 
-	apiServices, err := c.getAPIServicesToManageFn()
+	enabledApiServices, disabledApiServices, err := c.getAPIServicesToManageFn()
 	if err != nil {
 		return err
 	}
-	ready, err := c.precondition(apiServices)
+
+	syncDisabledAPIServicesErr := c.syncDisabledAPIServices(ctx, disabledApiServices)
+	conditionAPIServicesDegraded := operatorv1.OperatorCondition{
+		Type:   "APIServicesDegraded",
+		Status: operatorv1.ConditionFalse,
+	}
+	if syncDisabledAPIServicesErr != nil {
+		conditionAPIServicesDegraded.Status = operatorv1.ConditionTrue
+		conditionAPIServicesDegraded.Reason = "DisabledAPIServicesPresent"
+		conditionAPIServicesDegraded.Message = syncDisabledAPIServicesErr.Error()
+	}
+
+	ready, err := c.preconditionForEnabledAPIServices(enabledApiServices)
 	if err != nil {
 		if _, _, updateErr := v1helpers.UpdateStatus(ctx, c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
 			Type:    "APIServicesAvailable",
 			Status:  operatorv1.ConditionFalse,
 			Reason:  "ErrorCheckingPrecondition",
 			Message: err.Error(),
-		})); updateErr != nil {
-			return errors.NewAggregate([]error{err, updateErr})
+		}), v1helpers.UpdateConditionFn(conditionAPIServicesDegraded)); updateErr != nil {
```

Comment:
I don't want to complicate the code but it looks like it might be easy to forget which conditions should be updated. I like the way it was done in the workload controller (https://github.com/openshift/library-go/blob/c021d427e60e6cbd278412a5a5e49784f2c2144b/pkg/operator/apiserver/controller/workload/workload.go#L182). Maybe we could do something similar here?

I think it boils down to:
1. define conditions with default values
2. register an update function for conditions in the defer statement
3. when you need to end processing, just set the condition and return  (it will be updated in the defer statement).

Reply:
Done. I did not find any use for `updateGenerationFn` though.